File: src/code-samples/langchain/message-serialization.py
# :snippet-start: message-serialization-py
from langchain.messages import HumanMessage
from langchain_core.load import dumpd, load

message = HumanMessage("What is the capital of France?")

# Serialize to a plain dict
serialized = dumpd(message)

# Deserialize back to a message object
restored = load(serialized)
# :snippet-end:

# :remove-start:
if __name__ == "__main__":
    assert isinstance(serialized, dict)
    assert isinstance(restored, HumanMessage)
    assert restored.content == message.content
    print("✓ Message serialization round-trip works")
# :remove-end:

File: src/code-samples/langchain/message-serialization.ts
// :snippet-start: message-serialization-js
import { HumanMessage } from "@langchain/core/messages";
import { load } from "@langchain/core/load";

const message = new HumanMessage("What is the capital of France?");

// Serialize to a plain object
const serialized = message.toJSON();

// Deserialize back to a message object
const restored = await load<HumanMessage>(JSON.stringify(serialized));
// :snippet-end:

// :remove-start:
async function main() {
  if (typeof serialized !== "object" || serialized === null) {
    throw new Error(`Expected plain object, got ${typeof serialized}`);
  }
  if (!(restored instanceof HumanMessage)) {
    throw new Error(`Expected HumanMessage, got ${restored?.constructor?.name}`);
  }
  if (restored.content !== message.content) {
    throw new Error(
      `Expected content "${message.content}", got "${restored.content}"`,
    );
  }
  console.log("✓ Message serialization round-trip works");
}
main();
// :remove-end:
